model/shared/src/main/scala/clue/model/json/package.scala

```diff
@@ -0,0 +1,202 @@
+    ).reduceLeft(_ or _)
+
+
+  // ---- FromServer
+
+  import StreamingMessage.FromServer._
+
+  implicit val EncoderConnectionError: Encoder[ConnectionError] =
+    (a: ConnectionError) =>
+      Json.obj(
+        "type"    -> Json.fromString("connection_error"),
+        "payload" -> a.payload
+      )
+
+  implicit val DecoderConnectionError: Decoder[ConnectionError] =
+    (c: HCursor) =>
+      for {
+        _ <- checkType(c, "connection_error")
+        p <- c.downField("payload").as[Json]
+      } yield ConnectionError(p)
+
+  implicit val EncoderDataWrapper: Encoder[DataWrapper] =
+    (a: DataWrapper) =>
+      Json.obj(
+        "data" -> a.data
+      )
+
+  implicit val DecoderDataWrapper: Decoder[DataWrapper] =
+    (c: HCursor) =>
+      c.downField("data").as[Json].map(DataWrapper(_))
+
+  implicit val EncoderData: Encoder[Data] =
+    (a: Data) =>
+      Json.obj(
+        "type"    -> Json.fromString("data"),
+        "id"      -> Json.fromString(a.id),
+        "payload" -> a.payload.asJson
+      )
+
+  implicit val DecoderData: Decoder[Data] =
+    (c: HCursor) =>
+      for {
+        _ <- checkType(c, "data")
+        i <- c.downField("id").as[String]
+        p <- c.downField("payload").as[DataWrapper]
+      } yield Data(i, p)
+
+  implicit val EncoderError: Encoder[Error] =
+    (a: Error) =>
+      Json.obj(
+        "type"    -> Json.fromString("error"),
+        "id"      -> Json.fromString(a.id),
+        "payload" -> a.payload
+      )
+
+  implicit val DecoderError: Decoder[Error] =
+    (c: HCursor) =>
+      for {
+        _ <- checkType(c, "error")
+        i <- c.downField("id").as[String]
+        p <- c.downField("payload").as[Json]
+      } yield Error(i, p)
+
+  implicit val EncoderComplete: Encoder[Complete] =
+    (a: Complete) =>
+      Json.obj(
+        "type"    -> Json.fromString("complete"),
+        "id"      -> Json.fromString(a.id)
+      )
+
+  implicit val DecoderComplete: Decoder[Complete] =
+    (c: HCursor) =>
+      for {
+        _ <- checkType(c, "complete")
+        i <- c.downField("id").as[String]
+      } yield Complete(i)
+
+
+  implicit val EncoderFromServer: Encoder[StreamingMessage.FromServer] =
+    Encoder.instance {
+      case _ @ ConnectionAck       => encodeCaseObject("connection_ack")
+      case m @ ConnectionError(_)  => m.asJson
+      case _ @ ConnectionKeepAlive => encodeCaseObject("ka")
+      case m @ Data(_, _)          => m.asJson
+      case m @ Error(_, _)         => m.asJson
+      case m @ Complete(_)         => m.asJson
+    }
+
+  implicit val DecoderFromServer: Decoder[StreamingMessage.FromServer] =
+    List[Decoder[StreamingMessage.FromServer]](
+      decodeCaseObject("connection_ack", ConnectionAck),
+      Decoder[ConnectionError].widen,
+      decodeCaseObject("ka", ConnectionKeepAlive),
+      Decoder[Data].widen,
+      Decoder[Error].widen,
+      Decoder[Complete].widen
+    ).reduceLeft(_ or _)
```

Comment:
If decoding fails, will the `DecodingFailure` inform of all the decoders that were attempted (via `c.history` in `checkType`), or will it just report the last one?